integrationTests/state/stateTrie_test.go

```diff
@@ -1171,3 +1175,166 @@ func collapseTrie(state state.AccountHandler, t *testing.T) {
+	time.Sleep(time.Second * 5)
+
+	round, nonce = integrationTests.ProposeAndSyncOneBlock(t, nodes, idxProposers, round, nonce)
+	time.Sleep(time.Second * 5)
+
+	assert.Equal(t, uint64(2), nodes[0].BlockChain.GetCurrentBlockHeader().GetNonce())
+	assert.Equal(t, uint64(2), nodes[1].BlockChain.GetCurrentBlockHeader().GetNonce())
+
+	shardIdToRollbackLastBlock := uint32(0)
+	integrationTests.ForkChoiceOneBlock(nodes, shardIdToRollbackLastBlock)
+	integrationTests.ResetHighestProbableNonce(nodes, shardIdToRollbackLastBlock, 1)
+	integrationTests.EmptyDataPools(nodes, shardIdToRollbackLastBlock)
+
+	assert.Equal(t, uint64(1), nodes[0].BlockChain.GetCurrentBlockHeader().GetNonce())
+	assert.Equal(t, uint64(2), nodes[1].BlockChain.GetCurrentBlockHeader().GetNonce())
+
+	nonces := []*uint64{new(uint64), new(uint64)}
+	atomic.AddUint64(nonces[0], 2)
+	atomic.AddUint64(nonces[1], 3)
+
+	numOfRounds := 6
+	integrationTests.ProposeBlocks(
+		nodes,
+		&round,
+		idxProposers,
+		nonces,
+		numOfRounds,
+	)
+
+	err := shardNode.AccntState.RecreateTrie(rootHash)
+	assert.Nil(t, err)
+	assert.Equal(t, uint64(7), nodes[0].BlockChain.GetCurrentBlockHeader().GetNonce())
+	assert.Equal(t, uint64(8), nodes[1].BlockChain.GetCurrentBlockHeader().GetNonce())
+}
+
+func TestTriePruningWhenBlockIsFinal(t *testing.T) {
+	if testing.Short() {
+		t.Skip("this is not a short test")
+	}
+
+	fmt.Println("Setup nodes...")
+	numOfShards := 1
+	nodesPerShard := 1
+	numMetachainNodes := 1
+
+	idxProposers := []int{0, 1}
+	senderShard := uint32(0)
+	round := uint64(0)
+	nonce := uint64(0)
+
+	valMinting := big.NewInt(100)
+	valToTransferPerTx := big.NewInt(2)
+
+	advertiser := integrationTests.CreateMessengerWithKadDht(context.Background(), "")
+	_ = advertiser.Bootstrap()
+
+	nodes := integrationTests.CreateNodes(
+		numOfShards,
+		nodesPerShard,
+		numMetachainNodes,
+		integrationTests.GetConnectableAddress(advertiser),
+	)
+	integrationTests.DisplayAndStartNodes(nodes)
+
+	defer func() {
+		_ = advertiser.Close()
+		for _, n := range nodes {
+			_ = n.Node.Stop()
+		}
+	}()
+
+	fmt.Println("Generating private keys for senders and receivers...")
+	generateCoordinator, _ := sharding.NewMultiShardCoordinator(uint32(numOfShards), 0)
+	nrTxs := 20
+
+	//sender shard keys, receivers  keys
+	sendersPrivateKeys := make([]crypto.PrivateKey, nrTxs)
+	receiversPublicKeys := make(map[uint32][]crypto.PublicKey)
+	for i := 0; i < nrTxs; i++ {
+		sendersPrivateKeys[i], _, _ = integrationTests.GenerateSkAndPkInShard(generateCoordinator, senderShard)
+		_, pk, _ := integrationTests.GenerateSkAndPkInShard(generateCoordinator, senderShard)
+		receiversPublicKeys[senderShard] = append(receiversPublicKeys[senderShard], pk)
+	}
+
+	fmt.Println("Minting sender addresses...")
+	integrationTests.CreateMintingForSenders(nodes, senderShard, sendersPrivateKeys, valMinting)
+
+	shardNode := nodes[0]
+	rootHash, _ := shardNode.AccntState.RootHash()
```

Comment:
Also here, you take the root hash of the genesis block

Reply:
Changed.